climada/engine/supplychain.py

```diff
@@ -0,0 +1,395 @@
+"""
+This file is part of CLIMADA.
+
+Copyright (C) 2017 ETH Zurich, CLIMADA contributors listed in AUTHORS.
+
+CLIMADA is free software: you can redistribute it and/or modify it under the
+terms of the GNU Lesser General Public License as published by the Free
+Software Foundation, version 3.
+
+CLIMADA is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License along
+with CLIMADA. If not, see <https://www.gnu.org/licenses/>.
+
+---
+
+Define the SupplyChain class.
+"""
+
+__all__ = ['SupplyChain']
+
+import os
+import logging
+import datetime as dt
+from pathlib import Path
+from tqdm import tqdm
+import numpy as np
+import pandas as pd
+from iso3166 import countries_by_alpha3
+from iso3166 import countries_by_numeric
+from climada.util.files_handler import download_file
+from climada.engine import Impact
+from climada.util.constants import SYSTEM_DIR
+from climada.entity.exposures.base import Exposures
+
+LOGGER = logging.getLogger(__name__)
+WIOD_FILE_LINK = 'http://www.wiod.org/protected3/data16/wiot_ROW/'
+
+class SupplyChain():
+    """SupplyChain definition. It provides methods for the entire supplychain-risk
+    workflow and attributes holding the workflow's data and results.
+
+    Attributes:
+        mriot_data (np.array): 2-dim np.array of floats representing the data of
+            a full multi-regional input-output table (mriot).
+        countries (np.array): 1-dim np.array of strings containing the full list
+            of countries represented in the mriot, corresponding to the columns/
+            rows of mriot_data. For these countries risk calculations can be made.
+        countries_iso3 (np.array): similar to .countries, but containing the
+            countries' respective iso3-codes.
+        sectors (np.array): 1-dim np.array of strings containing the full
+            list of sectors represented in the mriot, corresponding to the columns/
```

Comment:
Both countries and sectors are in both rows/columns. 

Columns and rows of the WIOT table have countries, and for each country, values for each sector. So the num of cols/rows = n_countries x n_sectors

Reply:
That info is now been removed all together as whoever uses this package must know how the WIOD table is structured

Reply:
In what commit? Please refer to the changes so that one can see what was done. I am not yet convinced by your argument here.

Reply:
the changes were made in commit 4d04eaca49dd93bc711f487be24c2b6dabe8f9c5

If you are up to use this module, which uses WIOD IO tables at its core, I think we can agree one needs to know what this tables are, and how they are structured; then it follows one knows what is on the cols, what is on the rows therefore there is not need to repeat it in the docstring.

Reply:
That is a fair point. However, it makes it rather difficult to review and/or debug in the future, in particular the parts where indices cuts are made in the code which are not obvious. But, you can leave it as is, I guess it is a matter of taste. Thanks for linking to the commit!

Reply:
The only hardcoded bit are lines 131-135:
```
        col_sectors = 1
        col_iso3 = 2
        end_row_sectors = 61
        start_row, end_row = (5, 2469)
        start_col, end_col = (4, 2468)
```
I think for debugging one must know what is the data structure he/she is dealing with. There are also info in the tutorial (which could be improved indeed) regarding data structure etc...

Reply:
Sorry for the confusion, the hard-coded indices have nothing to do with that part, you are correct. My mistake. 
However, saying what the WIOT table is made of is useful I think, since it is not necessary to use WIOD tables in this module, but any WIOT table is possible.